packages/aws-cdk/test/api/hotswap/s3-bucket-hotswap-deployments.test.ts

```diff
@@ -0,0 +1,863 @@
+  // GIVEN
+  setup.setCurrentCfnStackTemplate({
+    Resources: {
+      S3Deployment: {
+        Type: 'Custom::NotCDKBucketDeployment',
+        Properties: {
+          ServiceToken: 'a-lambda-arn',
+          SourceBucketNames: [
+            'src-bucket',
+          ],
+          SourceObjectKeys: [
+            'src-key-old',
+          ],
+          DestinationBucketName: 'dest-bucket',
+        },
+      },
+    },
+  });
+  const cdkStackArtifact = setup.cdkStackArtifactOf({
+    template: {
+      Resources: {
+        S3Deployment: {
+          Type: 'Custom::NotCDKBucketDeployment',
+          Properties: {
+            ServiceToken: 'a-lambda-arn',
+            SourceBucketNames: [
+              'src-bucket',
+            ],
+            SourceObjectKeys: [
+              'src-key-new',
+            ],
+            DestinationBucketName: 'dest-bucket',
+          },
+        },
+      },
+    },
+  });
+
+  // WHEN
+  const deployStackResult = await cfnMockProvider.tryHotswapDeployment(cdkStackArtifact);
+
+  // THEN
+  expect(deployStackResult).toBeUndefined();
+  expect(mockLambdaInvoke).not.toHaveBeenCalled();
+});
+
+test('can use the DestinationBucketKeyPrefix property', async () => {
+  // GIVEN
+  setup.setCurrentCfnStackTemplate({
+    Resources: {
+      S3Deployment: {
+        Type: 'Custom::CDKBucketDeployment',
+        Properties: {
+          ServiceToken: 'a-lambda-arn',
+          SourceBucketNames: [
+            'src-bucket',
+          ],
+          SourceObjectKeys: [
+            'src-key-old',
+          ],
+          DestinationBucketName: 'dest-bucket',
+          DestinationBucketKeyPrefix: 'my-key/some-old-prefix',
+        },
+      },
+    },
+  });
+  const cdkStackArtifact = setup.cdkStackArtifactOf({
+    template: {
+      Resources: {
+        S3Deployment: {
+          Type: 'Custom::CDKBucketDeployment',
+          Properties: {
+            ServiceToken: 'a-lambda-arn',
+            SourceBucketNames: [
+              'src-bucket',
+            ],
+            SourceObjectKeys: [
+              'src-key-new',
+            ],
+            DestinationBucketName: 'dest-bucket',
+            DestinationBucketKeyPrefix: 'my-key/some-new-prefix',
+          },
+        },
+      },
+    },
+  });
+
+  // WHEN
+  const deployStackResult = await cfnMockProvider.tryHotswapDeployment(cdkStackArtifact);
+
+  // THEN
+  expect(deployStackResult).not.toBeUndefined();
+  payload.ResourceProperties = {
+    SourceBucketNames: [
+      'src-bucket',
+    ],
+    SourceObjectKeys: [
+      'src-key-new',
+    ],
+    DestinationBucketName: 'dest-bucket',
+    DestinationBucketKeyPrefix: 'my-key/some-new-prefix',
+  };
+
+  expect(mockLambdaInvoke).toHaveBeenCalledWith({
+    FunctionName: 'a-lambda-arn',
+    Payload: JSON.stringify(payload),
+  });
+});
+
```

Comment:
I feel like many of the above tests are redundant.

1. No reason to check for the new resource case again, it's already covered.
2. No reason to check the missing property case again, it's already covered.
3. No reason to signal out `DestinationBucketKeyPrefix` especially, just include it in other tests (this way, we would need a separate test for each possible property).

In general: be cognizant of the length of this file. Do not introduce tests that do not affect the coverage of the production code you've written for S3 deployments, but just re-affirm existing common code is working.